doc/reference/misc/data_structures.rst

```diff
@@ -0,0 +1,443 @@
+indicate what user data is "owned" by that node.  Instead, the
+expectation is that the node will be itself embedded within a
+user-defined struct.  Macros are provided to retrieve a user struct
+address from the embedded node pointer in a clean way.  The purpose
+behind this design is to allow the collections to be used in contexts
+where dynamic allocation is disallowed (i.e. there is no need to
+allocate node objects because the memory is provided by the user).
+
+Note also that these libraries are uniformly unsynchronized; access to
+them is not threadsafe by default.  These are data structures, not
+synchronization primitives.  The expectation is that any locking
+needed will be provided by the user.
+
+Single-linked List
+==================
+
+Zephyr provides a ``sys_slist_t`` type for storing simple
+singly-linked list data (i.e. data where each list element stores a
+pointer to the next element, but not the previous one).  This supports
+constant-time access to the first (head) and last (tail) elements of
+the list, insertion before the head and after the tail of the list and
+constant time removal of the head.  Removal of subsequent nodes
+requires access to the "previous" pointer and thus can only be
+performed in linear time by searching the list.
+
+The ``sys_slist_t`` struct may be instantiated by the user in any
+accessible memory.  It should be initialized with either
+``sys_slit_init()`` or by static assignment from SYS_SLIST_STATIC_INIT
+before use.  Its interior fields are opaque and should not be accessed
+by user code.
+
+The end nodes of a list may be retrieved with
+``sys_slist_peek_head()`` and ``sys_slist_peek_tail()``, which will
+return NULL if the list is empty, otherwise a pointer to a
+``sys_snode_t`` struct.
+
+The ``sys_snode_t`` struct represents the data to be inserted.  In
+general, it is expected to be allocated/controlled by the user,
+usually embedded within a struct which is to be added tracked by the
+list.  The container struct pointer may be retrieved from a list node
+using ``SYS_SLIST_CONTAINER()``, passing it the struct name of the
+containing struct and the field name of the node.  Internally, the
+``sys_snode_t`` struct contains only a next pointer, which may be
+accessed with ``sys_slist_peek_next()``.
+
+Lists may be modified by adding a single node at the head or tail with
+``sys_slist_prepend()`` and ``sys_slist_append()``.  They may also
+have a node added to an interior point with ``sys_slist_insert()``,
+which inserts a new node after an existing one.  Similarly
+``sys_slist_remove()`` will remove a node given an pointer to its
+predecessor.  These operations are all constant time.
+
+Convenience routines exist for more complicated modifications to a
+list.  ``sys_slist_merge_slist()`` will append an entire list to an
+existing one.  ``sys_slist_append_list()`` will append a bounded
+subset of an existing list in constant time.  And
```

Comment:
Maybe "foo will ..., bar will ..., and baz will ...", to avoid "and" at start of sentence.